src/mem/arena.rs

```diff
@@ -8,18 +8,54 @@
 
 use core::cell::Cell;
 use core::marker::PhantomData;
+use core::mem;
 use core::ptr::NonNull;
 use core::slice;
 
 use static_assertions::assert_obj_safe;
 
+use zerocopy::AsBytes;
+use zerocopy::FromBytes;
+use zerocopy::LayoutVerified;
+
+use crate::mem::align_to;
+
 /// An error indicating that an [`Arena`] has run out of allocatable
-/// memory.
+/// memory or that memory that is more aligned than is supported
+/// was requested.
+///
+/// Additionally, `OutOfMemory` implements [`Arena`] itself, acting as the
+/// "trivial" no-memory-was-provided arena. This is particularly useful for
+/// when it is known that no memory will ever be allocated on the arena.
 ///
 /// [`Arena`]: trait.Arena.html
 #[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
 pub struct OutOfMemory;
 
+unsafe impl Arena for OutOfMemory {
+    fn alloc_aligned(
+        &self,
+        len: usize,
+        align: usize,
+    ) -> Result<&mut [u8], OutOfMemory> {
+        assert!(align.is_power_of_two());
+        if len == 0 && align <= mem::align_of::<u128>() {
```

Comment:
Why limit this to `align <= mem::align_of::<u128>()`? As far as I can tell, the following implementation should be fine:

```rust
assert!(align.is_power_of_two());
if len == 0 {
    Ok(unsafe { &mut *(align as *mut u8) })
} else {
    Err(*self)
}
```

Reply:
Excellent observation!